src/main/java/com/google/sps/servlets/MapServlet.java

```diff
@@ -0,0 +1,129 @@
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.gson.Gson;
+import com.google.sps.Trip;
+import com.google.sps.TripDay;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/get-map")
+public class MapServlet extends HttpServlet {
+
+  private final String TRIP_KEY_PARAM = "tripKey";
+
+  /**
+   * Checks for invalid cases (no user or tripKey).
+   * Gets the locations from datastore and prints them to writer.
+   */  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) 
+      throws IOException {
+    response.setContentType("application/json;");
+
+    // Initialize Datastore.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get current user.
+    Entity userEntity = AuthServlet.getCurrentUserEntity();
+
+    // Get tripKey.
+    String stringTripKey = request.getParameter(TRIP_KEY_PARAM);
+
+    // if no tripKey return to trips page.
+    if (stringTripKey == null) {
+      response.getWriter().println("No trip Key");
+      response.sendRedirect("/trips/");
+  
+    // if no user Entity return home.
+    } else if (userEntity == null) {
+      response.getWriter().println("No current User");
+      response.sendRedirect("/");
+    } else { 
+      Key tripKey = KeyFactory.stringToKey(stringTripKey);
+
+      // Gets the locations from datastore and writes them to .../get-map
+      doGetMap(response, datastore, userEntity, tripKey);
+    }
+  }
+
+  /**
+   * Gets the locations and prints them to writer.
+   */
+  public void doGetMap(HttpServletResponse response, 
+      DatastoreService datastore, Entity userEntity, Key tripEntityKey) throws IOException {
+  
+    // Get trip Entity based on trip key.
+    // Trip entity is needed to ensure that tripKey corresponds to a Trip under the current user.
+    Filter tripKeyFilter =
+      new FilterPredicate("__key__", FilterOperator.EQUAL, tripEntityKey);
+    Query tripQuery = new Query(Trip.TRIP, userEntity.getKey());
+    tripQuery.setFilter(tripKeyFilter);
+    PreparedQuery tripResults = datastore.prepare(tripQuery);
+    Entity tripEntity = tripResults.asSingleEntity();
+
+    // If no trip is found then redirect home
+    if (tripEntity == null) {
+      response.getWriter().println("No trip found");
+      response.sendRedirect("/");
+    } else {
```

Comment:
Thanks for the comment, I agree! I changed the helper method `doGetMap` so that it returns a String. In the error case, it returns "No trip found" and otherwise it returns the JSON string. I then write the returned String to the writer in the `doGet` method. I have also modified my tests to test the functionality of both `doGetMap` and `doGet`.